Dear Colleague,

We are pleased to submit our manuscript "Phylotrack: C++ and Python libraries for *in silico* phylogenetic tracking" for consideration as an Application Note in Bioinformatics.

Phylotrack is an efficient, flexible software library for tracking and analyzing phylogenies in computational agent-based evolutionary simulations.
It fills an important gap by offering a general-purpose solution for phylogeny tracking within end-users' own simulation code.
Key features include:

- efficient C++ implementation with Python bindings;
- flexible configuration of operational taxonomic units,
- configurable memory-saving pruning of extinct lineages,
- *in situ* calculation of common phylogenetic topology statistics,
- saving and loading phylogeny records.

Our manuscript describes the implementation and functionality of Phylotrack, benchmarks its runtime performance, and discusses its applications and relationship to existing tools in the field.
We believe Phylotrack will be a valuable resource for researchers using *in silico* evolution to study evolutionary dynamics and generate synthetic data for bioinformatics research.

The software is freely available under an open source license at <https://github.com/emilydolson/phylotrackpy>, with source code, documentation, and precompiled packages provided.
It is also installable via PyPI.

We appreciate your consideration of our manuscript and look forward to your response.

Sincerely,

Emily Dolson, Santiago Rodriguez-Papa, and Matthew Andres Moreno